### PR TITLE
🎉🐛: Source mongoDB: implement building JsonSchema with 'properties' for fields with type 'object'

### LINKED ISSUE #8586: 🐛 Destination BigQuery denormalized (MongoDB source): JSON table encountered too many errors, giving up.
## Environment
- **Airbyte version**:  0.33.5-alpha
- **OS Version / Instance**: Container-Optimized OS from Google 
- **Deployment**: Kubernetes
- **Source Connector and version**: mongo v2 0.1.8
- **Destination Connector and version**: BigQuery Denormalized 0.1.10
- **Severity**: High 
- **Step where error happened**: Sync job

## Current Behavior
Sync fails with the error:
```
[35mdestination[0m - 2021-12-07 14:23:12 INFO DefaultAirbyteStreamFactory(lambda$create$0):61 - 2021-12-07 14:23:12 [32mINFO[m i.a.i.d.b.BigQueryRecordConsumer(closeNormalBigqueryStreams):312 - {} - Removing tmp tables...
[35mdestination[0m - 2021-12-07 14:23:12 INFO DefaultAirbyteStreamFactory(lambda$create$0):61 - 2021-12-07 14:23:12 [32mINFO[m i.a.i.d.b.BigQueryRecordConsumer(closeNormalBigqueryStreams):314 - {} - Finishing destination process...completed
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - Exception in thread "main" java.lang.RuntimeException: com.google.cloud.bigquery.BigQueryException: Error while reading data, error message: JSON table encountered too many errors, giving up. Rows: 1; errors: 1. Please look into the errors[] collection for more details.
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryRecordConsumer.lambda$closeNormalBigqueryStreams$6(BigQueryRecordConsumer.java:289)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.commons.lang.Exceptions.castCheckedToRuntime(Exceptions.java:52)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.commons.lang.Exceptions.toRuntime(Exceptions.java:39)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryRecordConsumer.lambda$closeNormalBigqueryStreams$7(BigQueryRecordConsumer.java:279)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at java.base/java.util.HashMap$Values.forEach(HashMap.java:1068)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryRecordConsumer.closeNormalBigqueryStreams(BigQueryRecordConsumer.java:279)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryRecordConsumer.close(BigQueryRecordConsumer.java:169)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryDenormalizedRecordConsumer.close(BigQueryDenormalizedRecordConsumer.java:81)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.base.FailureTrackingAirbyteMessageConsumer.close(FailureTrackingAirbyteMessageConsumer.java:62)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.base.IntegrationRunner.consumeWriteStream(IntegrationRunner.java:152)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.base.IntegrationRunner.run(IntegrationRunner.java:128)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryDenormalizedDestination.main(BigQueryDenormalizedDestination.java:209)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - Caused by: com.google.cloud.bigquery.BigQueryException: Error while reading data, error message: JSON table encountered too many errors, giving up. Rows: 1; errors: 1. Please look into the errors[] collection for more details.
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at com.google.cloud.bigquery.Job.reload(Job.java:411)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at com.google.cloud.bigquery.Job.waitFor(Job.java:248)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	at io.airbyte.integrations.destination.bigquery.BigQueryRecordConsumer.lambda$closeNormalBigqueryStreams$6(BigQueryRecordConsumer.java:282)
[35mdestination[0m - 2021-12-07 14:23:12 ERROR LineGobbler(voidCall):82 - 	... 11 more
2021-12-07 14:23:12 INFO KubePodProcess(getReturnCode):583 - Exit code for pod destination-bigquery-denormalized-sync-27-0-xpmjg is 0
2021-12-07 14:23:12 INFO KubePodProcess(exitValue):598 - Closed all resources for pod destination-bigquery-denormalized-sync-27-0-xpmjg
2021-12-07 14:23:12 INFO DefaultReplicationWorker(run):141 - Destination thread complete.

```

yet the sync is marked as _Succeeded_ with no data in destination.

## Expected Behavior
1. Sync is not marked as succeeded when it failed.
2. Bug is fixed or the error message would explain what is wrong with the source data.

## Steps to Reproduce
1. set-up connectors
2. run sync

## Are you willing to submit a PR?
no

### PR DESCRIPTION
**Pull request by yurii-bidiuk**
## What
Fixed #8586
On the source-mongodb-v2 implemented building JsonSchema with nested properties for fields with type "object"

## How
Add recursive fetching of fields and types from mongo collection. Modified `CommonField` and `Field` classes to support subfields.
 

## Recommended reading order
1. `MongoUtils.java`
2. `MongoDBSource.java`
3. `CatalogHelpers.java`

## 🚨 User Impact 🚨
Users will be able to see on the UI all fields and types including all sublevels fields.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Review by alexandr-shegeda — APPROVED** [review]
LGTM

File: airbyte-db/lib/src/main/java/io/airbyte/db/mongodb/MongoUtils.java [hunk 1]
```diff
@@ -188,21 +199,52 @@ private static ObjectNode readField(final BsonReader reader,
...
+      var fieldNode = new TreeNode<>(new CommonField<>(transformName(types, key), type));
+      if (type.equals(DOCUMENT)) {
+        setSubFields(collection, fieldNode, key);
+      }
+      return fieldNode;
+    }).toList();
+  }
+
+  /**
+   * If one field has different types in 2 and more documents, the name is transofmed to
```
Comment (noahkawasaki-airbyte) [inline airbyte-db/lib/src/main/java/io/airbyte/db/mongodb/MongoUtils.java:217]: Small typo: transformed

**Review by noahkawasaki-airbyte — APPROVED** [review]
LGTM too, did you run tests in CI? I dont see them having run from "Sonar Scan / Tests for ${{ matrix.module }} (pull_request)". Also are you going to bump version in this PR or do you usually do a follow up?

File: airbyte-db/lib/src/main/java/io/airbyte/db/mongodb/MongoUtils.java (hunk 1 above)
Comment (yurii-bidiuk) [reply airbyte-db/lib/src/main/java/io/airbyte/db/mongodb/MongoUtils.java]: fixed

**Comment by yurii-bidiuk** [comment]
> LGTM too, did you run tests in CI? I dont see them having run from "Sonar Scan / Tests for ${{ matrix.module }} (pull_request)". Also are you going to bump version in this PR or do you usually do a follow up?

@noahkawasaki-airbyte thanks for the review. 
Tests are passed in CI and the version is bumped in the new commit. Going to publish

**Pull request merged**